### DISCUSSION TITLE
Finding an API / class to query the Another iModel from frontend

### DISCUSSION THREAD

**Discussion by Adu2120** [iTwin/itwinjs-core#4317] [Q&A] [answered]
I am trying to query the another iModel into our iTwin-viewer app. I have to fetch the iModel related data which is inside the IModelDb. So, Is there any API available where we can retrieve that data? 
I checked this sample: [Sample 1](https://www.itwinjs.org/sandboxes/iTwinPlatform/iModels%20OData). But the ReportingClient is not working in the newer versions of '@itwin/insights-client'.
I tried [Select Elements in a particular Model](https://www.itwinjs.org/learning/backend/ecsql-queries/). But in this I didn't get that how to create an instance of IModelDb in frontend

Please help me to find this out

**Comment by ayushibasaiwal** [comment]
@pmconne @aruniverse Can you please look into this?

**Comment by aruniverse** [comment, marked as answer]
Hi, few questions

>  query the another iModel into our iTwin-viewer app

So you have iModel A opend in your viewer, and you want to query against A, or a different iModel B not being rendered by the viewer? What is it you are trying to query?

> But the ReportingClient is not working in the newer versions of '@itwin/insights-client'.

What is not working? What version of @itwin/insights-client worked and what version doesn't work?

> an instance of IModelDb in frontend

[IModelDb](https://www.itwinjs.org/reference/core-backend/imodels/imodeldb/) is a backend only class

If you want to query an existing open imodel on the frontend, you can use [IModelConnection.query](https://www.itwinjs.org/reference/core-frontend/imodelconnection/imodelconnection/query/).

**Reply by Adu2120** [reply]
> So you have iModel A opend in your viewer, and you want to query against A, or a different iModel B not being rendered by the viewer? What is it you are trying to query?

I want to query the iModel B into iModel A. 
1. Is it possible in front end? 
2. Instead of calling it in frontend If I create an API which takes the iModelId, iTwinId, query as a request parameters which gives us the response as the Result in JSON format. Is it possible? Because I read all the documentation in which they mentioned that they need the key or the bim file. Is there any way to query using above parameters?

**Reply by Adu2120** [reply]
I have small query, If there is an scenario where we have Project A in that we have 3 different iModels and if I want to query the any 2 iModels from the 3rd iModel. So, is it possible or not? In short, I want to query the iModel within project not the iModel which is present in another project.

**Reply by aruniverse** [reply]
Aditya, my apologies for not responding back earlier, mustve missed the notification.

Yes, what you're looking for is possible; you just need to:
1. Open an iModelConnection against that project/imodel you want to query
```ts
  const iModelConn = await CheckpointConnection.openRemote(iTwinId, iModelId);
```
3. Run your query
```ts
  const query =
    'SELECT ECInstanceId FROM Bis.GeometricModel3D WHERE IsPrivate = false AND IsTemplate = false';

  const iter = iModelConn.query(query, undefined, {
    rowFormat: QueryRowFormat.UseJsPropertyNames,
  });

  const rows = [];
  for await (const row of iter) {
    rows.push(row.id);
  }

  console.log(rows);
```

**Reply by Adu2120** [reply]
It's Working. Thanks!!
